test/util/traffic_utils.h

```diff
@@ -0,0 +1,174 @@
+#include "microtar.h"
+
+#include <valhalla/baldr/graphreader.h>
+#include <valhalla/baldr/traffictile.h>
+
+#include <boost/property_tree/ptree.hpp>
+
+#include <cmath>
+#include <sstream>
+#include <sys/mman.h>
+#include <sys/stat.h>
+
+using namespace valhalla;
+
+namespace valhalla_tests {
+namespace utils {
+
+/**
+ * Generate a new GraphReader that doesn't re-use a previously
+ * statically initizalized tile_extract member variable.
+ *
+ * Useful if you need to reload a tile extract within the same
+ * process
+ */
+std::shared_ptr<baldr::GraphReader>
+make_clean_graphreader(const boost::property_tree::ptree& mjolnir_conf) {
+
+  // Wrapper sub-class to allow replacing the statically initialized
+  // tile_extract member variable
+  struct ResettingGraphReader : baldr::GraphReader {
+    ResettingGraphReader(boost::property_tree::ptree pt) : GraphReader(pt) {
+      // Reset the statically initialized tile_extract_ member variable
+      tile_extract_.reset(new baldr::GraphReader::tile_extract_t(pt));
+    }
+  };
+  return std::make_shared<ResettingGraphReader>(mjolnir_conf);
+}
+
+/** Copy of raw header for use with sizeof() **/
+typedef struct {
+  char name[100];
+  char mode[8];
+  char owner[8];
+  char group[8];
+  char size[12];
+  char mtime[12];
+  char checksum[8];
+  char type;
+  char linkname[100];
+  char _padding[255];
+} mtar_raw_header_t_;
+
+/*************************************************************/
+// Creates an empty traffic file
+//
+// To actually customize the traffic data, use `customize_live_traffic_data`
+//
+/*************************************************************/
+void build_live_traffic_data(const boost::property_tree::ptree& config) {
+
+  std::string tile_dir = config.get<std::string>("mjolnir.tile_dir");
+  std::string traffic_extract = config.get<std::string>("mjolnir.traffic_extract");
+
+  // Begin by seeding the traffic file,
+  // per-edge customizations come in the step after
+  {
+    mtar_t tar;
+    auto tar_open_result = mtar_open(&tar, traffic_extract.c_str(), "w");
+    if (tar_open_result != MTAR_ESUCCESS) {
+      throw std::runtime_error("Could not create traffic tar file");
+    }
+
+    baldr::GraphReader reader(config.get_child("mjolnir"));
+
+    auto tile_ids = reader.GetTileSet();
+    // Traffic data works like this:
```

Comment:
Should this explanation and the utility around reading/writing traffic tar files/tiles be in the main libvalhalla?

Reply:
Perhaps the explanation could also live in the graphreader but the code itself here is not meant for production and should stay on the testing side of the code.